Netkan/Processors/QueueHandler.cs

```diff
@@ -0,0 +1,206 @@
+                return new Metadata(JObject.Parse(msg.Body));
+            }
+            log.Debug("No metadata in queue");
+            return null;
+        }
+
+        private Message getFromQueue(string url)
+        {
+            log.DebugFormat("Looking for message from {0}", url);
+            var resp = client.ReceiveMessage(new ReceiveMessageRequest()
+            {
+                QueueUrl            = url,
+                MaxNumberOfMessages = 1,
+                VisibilityTimeout   = (int)TimeSpan.FromMinutes(15).TotalSeconds,
+            });
+            var msg = resp.Messages.FirstOrDefault();
+            if (msg != null)
+            {
+                log.Debug("Message received");
+                client.DeleteMessage(new DeleteMessageRequest()
+                {
+                    QueueUrl      = url,
+                    ReceiptHandle = msg.ReceiptHandle,
+                });
+            }
+            return msg;
+        }
+
+        private void sendCkan(Metadata ckan, Metadata netkan, bool success, string err = null)
+        {
+            var attribs = new Dictionary<string, MessageAttributeValue>()
+            {
+                {
+                    "ModIdentifier",
+                    new MessageAttributeValue()
+                    {
+                        DataType    = "String",
+                        StringValue = netkan.Identifier
+                    }
+                },
+                {
+                    "Staged",
+                    new MessageAttributeValue()
+                    {
+                        DataType    = "String",
+                        StringValue = netkan.Staged.ToString()
+                    }
+                },
+                {
+                    "Success",
+                    new MessageAttributeValue()
+                    {
+                        DataType    = "String",
+                        StringValue = success.ToString()
+                    }
+                },
+                {
+                    "CheckTime",
+                    new MessageAttributeValue()
+                    {
+                        DataType    = "String",
+                        StringValue = DateTime.UtcNow.ToString("s", CultureInfo.InvariantCulture)
+                    }
+                },
+            };
+            if (!string.IsNullOrEmpty(err))
+            {
+                attribs.Add(
+                    "ErrorMessage",
+                    new MessageAttributeValue()
+                    {
+                        DataType    = "String",
+                        StringValue = err
+                    }
+                );
+            }            
+            
+            SendMessageRequest msg = new SendMessageRequest()
+            {
+                QueueUrl               = outputQueueURL,
+                MessageGroupId         = "1",
+                MessageDeduplicationId = Path.GetRandomFileName(),
```

Comment:
This ID is important to avoid duplicate messages hitting the queue. We could enable deduplication based on content (avoiding the need to generate this) which might be better? What I've done with the hack script is to MD5 the content before I send it.

Reply:
Yup, that's why I went with `Path.GetRandomFileName()`, "[a cryptographically strong, random string](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getrandomfilename)". I don't want the queue logic getting smart on us and deleting "duplicates" that aren't actually duplicates, because from what I can tell, we won't be generating any duplicates.

Reply:
It might be unlikely that we'll generate duplicates, but I feel we should trust the queue to do the RightThing&trade;. In dev it might be desirable to allow us to produce metadata over and over that is the same, but in production I think we should ensure that if we do send a duplicate message that our `MessageDeduplicationId` is generated from the body in a repeatable fashion.

Reply:
Well, we can't just hash the body (or use `ContentBasedDeduplication`), because all of the error messages will have a body of "{}". They'd be falsely marked as duplicates if multiple were sent within 5 minutes, and we'd potentially miss out an an error message for the status page.

The place to worry about duplicates for this container is upstream. The Inflator does the most CPU-, network-, and RAM-intensive work. If we've received and processed a duplicate message from `InboundDev.fifo,` we've already "lost'. Trying to generate dedup IDs at this point is shutting the barn door after the horse has escaped.

Reply:
That's a really fair point and in reality this will suit our use case. The client has retry mechanisms built in and there are scenarios where the message made it but the client didn't get the response the first.

And you're very right, the indexer will just go, nothing to see here and drop it. Thanks for persisting with my design stubbornness :-)
![Nothing-to-See-15a34a2fc727c8](https://user-images.githubusercontent.com/3467784/60062297-3203ab80-972b-11e9-8ada-fcf5c61ad8c8.jpg)